llvm/lib/Analysis/ValueTracking.cpp

```diff
@@ -705,28 +705,40 @@ static void computeKnownBitsFromCmp(const Value *V, CmpInst::Predicate Pred,
   }
 }
 
+static void computeKnownBitsFromCond(const Value *V, Value *Cond,
```

Comment:
Should we add a template helper to avoid duplicating the decomposition?
```
template<typename Callable>
void decomposeLogicalAndOr(Value *Cond, const Callable& Callback, unsigned Depth, bool Invert) {
Value *A, *B;
   if (Depth < MaxAnalysisRecursionDepth &&
       (Invert ? match(Cond, m_LogicalOr(m_Value(A), m_Value(B)))
               : match(Cond, m_LogicalAnd(m_Value(A), m_Value(B))))) {
    decomposeLogicalAndOr(A, Callback, Depth + 1, Invert);
    decomposeLogicalAndOr(B, Callback, Depth + 1, Invert);
    }
    // TODO: match logical not
    else
        Callback(Cond, Invert);
}

template<typename Callable>
void decomposeLogicalAndOrWithWorkList(Value *Cond, const Callable& Callback) {
    ...
}
```

Reply:
We could do something like https://github.com/llvm/llvm-project/commit/07f0dde1783a0ce7ea17bcf29438989e9077b310 to allow reuse of the full logic for other ValueTracking helpers, e.g. https://github.com/llvm/llvm-project/pull/80941. Seems to add a little overhead, but not much: http://llvm-compile-time-tracker.com/compare.php?from=9ac82f0d3ecf6c13669b0c7940920460c037a292&to=07f0dde1783a0ce7ea17bcf29438989e9077b310&stat=instructions:u